HDFS plugin: add replication_factor param

trakos opened the pull request:
When creating snapshots in HDFS repository, there was no way to choose a specific HDFS replication factor.
Setting `hdfs.dfs.replication` to any number is ignored on the client side in `FileContext` API.

Here's an excerpt from [FileContext.create documentation](https://hadoop.apache.org/docs/current/api/org/apache/hadoop/fs/FileContext.html#create-org.apache.hadoop.fs.Path-java.util.EnumSet-org.apache.hadoop.fs.Options.CreateOpts...-):

> The defaults for the following are SS [server-side] defaults of the file server implementing the target path.
> BufferSize - buffersize used in FSDataOutputStream
> Blocksize - block size for file blocks
> ReplicationFactor - replication for blocks
> ChecksumParam - Checksum parameters. server default is used if not specified.



HDFS allows setting a custom replication level when creating new files, but it has to be done via an argument.
This PR adds a new setting to HDFS repositories, `replication_factor`, which is then used whenever a file is created.
The default behavior is the same as it used to be - use HDFS cluster default.

Fixes https://github.com/elastic/elasticsearch/issues/90952 - I think the description there isn't accurate, as the setting is applied, it's just that this specific setting isn't effective on the HDFS client side.

Example API request to create an HDFS repository with replication factor set to 1:

```
PUT /_snapshot/estest
{
  "type": "hdfs",
  "settings": {
    "uri": "hdfs://namenode:9000/",
    "path": "/estest",
    "readonly": false,
    "replication_factor": 1,
    "conf.dfs.client.use.datanode.hostname": true,
    "conf.dfs.datanode.use.datanode.hostname": true
  }
}
```

In our use-case, the default replication factor on HDFS is too high for the amount of data that we want to snapshot. We might run out of disk space. It also isn't critical data for us, since most of it can be reindexed from elsewhere. Therefore, we want to use a lower replication setting. Changing it after snapshot creation isn't as viable, since we might run out of disk space first.

trakos committed:
HDFS plugin: add hdfs_replication_factor param

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

jbaiera commented on plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsBlobContainer.java:
```diff
@@ -230,14 +243,26 @@ private void writeToPath(
+            return baseOpts;
+        }
+
+        Options.CreateOpts[] newOpts = new Options.CreateOpts[baseOpts.length + 1];
```
Nit: It might be nice to have a pre-assembled array of CreateOpts that is reused all the time and only to do the copy dance if both that array and the base opts array have properties? Right now we do it all the time if the replication factor is set, even if we don't have to. I imagine it's not a ton of overhead, but it feels clunky?

jbaiera commented on plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsBlobContainer.java:
```diff
@@ -230,14 +243,26 @@ private void writeToPath(
         }
     }
 
+    private Options.CreateOpts[] getCreationOptions(Options.CreateOpts... baseOpts) {
```
Nit: The variable name `baseOpts` doesn't seem right. These are more like `overrideOpts` since they're provided on a call-by-call basis. The replication factor value is something I'd consider to be a base option since it's stable throughout the whole life of the object.

trakos replied:
I've removed that method altogether for now.

jbaiera commented on plugins/repository-hdfs/src/test/java/org/elasticsearch/repositories/hdfs/HdfsBlobStoreContainerTests.java:
```diff
@@ -145,7 +151,8 @@ public void testReadRange() throws Exception {
 
         // if not read only, directory will be created
-        hdfsBlobStore = new HdfsBlobStore(fileContext, "dir", 1024, false);
+        short replicationFactor = 8;
```
Can we break this verification out into its own separate test case?

jbaiera commented on docs/plugins/repository-hdfs.asciidoc:
```diff
@@ -77,6 +77,11 @@ include::repository-shared-settings.asciidoc[]
     the pattern with the hostname of the node at runtime (see
     link:repository-hdfs-security-runtime[Creating the Secure Repository]).
 
+`hdfs_replication_factor`::
```
Not _entirely_ sold on the property name - maybe something like `create_opts.replication`? That leaves the spec open to be extended in the future with other create opts if we want to, and also self-documents a bit better? I'm open to suggestions.

trakos replied:
Sure, I don't mind that. I've renamed it.

My only gripe is that it kind of feels like exposing internal implementation, since I assume that most users interact with HDFS using cli instead of Java client code. However, I assume that most users will have to check what it does in ES docs anyway.

jbaiera replied:
> most users interact with HDFS using cli instead of Java client code

That is a fair point - maybe just `replication_factor` then? Sorry for the back and forth 😄

trakos replied:
Sure! I've renamed it to `replication_factor`.

> Sorry for the back and forth 😄

No problem at all! It's much more complicated to modify after the fact, so it's good to consider all options now.

jbaiera reviewed (commented):
Thanks for the contribution, looking good so far! I had a couple comments, but nothing major.

jbaiera commented:
@elasticmachine test this please

trakos committed:
Fixes after review: rename param, add bounds validation, simplify create opts, split test

trakos commented on plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsRepository.java:
```diff
@@ -142,7 +144,14 @@ private HdfsBlobStore createBlobstore(URI blobstoreUri, String path, Settings re
+                bufferSize,
+                isReadOnly(),
+                haEnabled,
+                replicationFactor != null ? replicationFactor.shortValue() : null
```
Good point! I've added validation against `dfs.replication.min` and `dfs.replication.max`.

I've had to make the validation in `createBlobstore`, since that's where `hadoopConfiguration` is first created.
That makes the outside exception say `path is not accessible` when validation error occurs, which is a bit confusing:

```
{
  "error": {
    "root_cause": [
      {
        "type": "repository_exception",
        "reason": "[estest] Value of create_opts.replication [-2] must be >= dfs.replication.min [1]"
      }
    ],
    "type": "repository_verification_exception",
    "reason": "[estest] path  is not accessible on master node",
    "caused_by": {
      "type": "repository_exception",
      "reason": "[estest] Value of create_opts.replication [-2] must be >= dfs.replication.min [1]"
    }
  },
  "status": 500
}
```

Though it's also like that when you specify a wrong HDFS address, so I suppose that could be OK?

Alternatively, I could just hard-code a check for a 1 to 32767 range instead.

jbaiera replied:
Hmm, that error message seems to be tied to the verification logic in the `BlobStoreRepository` class. If it doesn't have a version of the repository yet when doing the verification, it creates it, but the creation is done inside of the try catch block that reports an error about not being able to write to a path. I think as long as you're doing the validation inside of the `createBlobstore` method, that is the right place. The ultimate error message is likely out of scope for this PR.

trakos commented:
Thank you for the thorough review, @jbaiera! I've updated the PR.

jbaiera commented on plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsBlobContainer.java:
```diff
@@ -230,14 +243,26 @@ private void writeToPath(
+
+        Options.CreateOpts[] newOpts = new Options.CreateOpts[baseOpts.length + 1];
+        System.arraycopy(baseOpts, 0, newOpts, 0, baseOpts.length);
+        newOpts[baseOpts.length] = Options.CreateOpts.repFac(replicationFactor);
```
I like the approach, but I think we should just do the array copy without the lang3 dependency. A helper method in the blob container would be fine!

trakos replied:
Makes sense! I've created a separate method `addOptionToArray`.

jbaiera commented on plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsRepository.java:
```diff
@@ -109,6 +111,30 @@ private HdfsBlobStore createBlobstore(URI blobstoreUri, String path, Settings re
+        if (replicationFactor != null && replicationFactor < minReplicationFactory) {
+            throw new RepositoryException(
+                metadata.name(),
+                MessageFormat.format(
+                    "Value of create_opts.replication [{0}] must be >= dfs.replication.min [{1}]",
+                    replicationFactor,
+                    minReplicationFactory
+                )
```
Fun fact: The RepositoryException extends ElasticsearchException which has message formatting baked in. You can just pass in the arguments as-is. I think the curly braces are fine with being empty as well, so just `{}` should work for the placeholders.

trakos replied:
Apologies, but I'm not entirely sure if I understand what's the suggested change here. The constructor with message formatting is in `ElasticsearchException`, but `RepositoryException` does not implement those from what I can see.

I've proposed adding vararg to `RepositoryException` constructor. Please let me know if it's OK to do it in this PR.

Alternatively, I can just instantiate `ElasticsearchException` instead, but then it will be wrapped in another `RepositoryException` [here](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java#L729), which makes the returned json error more complicated - root cause becomes `cannot create blob store` in that case.

jbaiera commented on plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsRepository.java:
```diff
@@ -142,7 +168,14 @@ private HdfsBlobStore createBlobstore(URI blobstoreUri, String path, Settings re
+                bufferSize,
+                isReadOnly(),
+                haEnabled,
+                replicationFactor != null ? replicationFactor.shortValue() : null
```
Thanks for adding the verification logic above, but if someone sets the dfs.replication.max higher than a short value, we'll still have an issue here - maybe we should add a third check above to make sure that the replicationFactor is between 0 and max short as a hard stop, regardless of the configured min and max?

trakos replied:
Good point! I've added a check that the replication factor is >= 1 and <= max short.

HDFS client lib throws `IllegalArgumentException` with message `Replication must be greater than 0` when it's below 1, but I suppose it's better to also check that ourselves so that we can give a better exception message.

jbaiera reviewed (commented):
Thanks again for further iterations on this, I think it's real close! Last bit of polish and we're all set!

trakos committed:
Fixes after review:

- Avoid using lang3.ArrayUtils, add `addOptionToArray` method instead
- Simplify test
- Rename `create_opts.replication` to `replication_factor`
- Make RepositoryException accept varargs arguments for message formatting
- Add validation that replication_factor is between 1 and 32767

trakos commented on plugins/repository-hdfs/src/test/java/org/elasticsearch/repositories/hdfs/HdfsBlobStoreContainerTests.java:
```diff
@@ -158,6 +164,36 @@ public void testReadRange() throws Exception {
+        writeBlob(container, "foo", new BytesArray(data), randomBoolean());
+        int pos = randomIntBetween(0, data.length / 2);
+        int len = randomIntBetween(pos, data.length) - pos;
+        assertArrayEquals(readBlobPartially(container, "foo", pos, len), Arrays.copyOfRange(data, pos, pos + len));
```
RIght, we don't! I've simplified this test further so that it's clearer what it tests.

trakos commented on plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsBlobContainer.java:
```diff
@@ -230,7 +247,9 @@ private void writeToPath(
         final byte[] buffer = new byte[blobSize < bufferSize ? Math.toIntExact(blobSize) : bufferSize];
-        try (FSDataOutputStream stream = fileContext.create(blobPath, createFlags, CreateOpts.bufferSize(buffer.length))) {
+
+        Options.CreateOpts[] createOptsWithBufferSize = ArrayUtils.add(createOpts, CreateOpts.bufferSize(buffer.length));
```
I've created a separate method `addOptionToArray` now.

trakos commented on server/src/main/java/org/elasticsearch/repositories/RepositoryException.java:
```diff
@@ -20,12 +21,12 @@
     }
 
-    public RepositoryException(String repository, String msg, Throwable cause) {
-        super("[" + (repository == null ? "_na" : repository) + "] " + msg, cause);
+    public RepositoryException(String repository, String msg, Throwable cause, Object... args) {
+        super("[" + (repository == null ? "_na" : repository) + "] " + LoggerMessageFormat.format(msg, args), cause);
```
Here's the change to `RepositoryException` that I mentioned in the other comment chain, which is based on a similar constructor in `ElasticsearchException` [here](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/ElasticsearchException.java#L142). Is that the desired change to avoid repeating `MessageFormat.format` calls?

trakos committed:
Add comments to RepositoryException constructors

trakos commented:
Thank you for taking a second look, @jbaiera! I've updated the PR.

Please let me know if you'd want me to [choose a different approach](https://github.com/elastic/elasticsearch/pull/94132#discussion_r1155163996) with the `RepositoryException` constructor change.

trakos commented:
Hey @jbaiera! Is this PR still on the radar? Can I somehow help to move it forward?

Thanks!

jbaiera commented:
Ah so sorry about that, this and the other PR seem to have fallen off my active review radar. I'll get new responses on here ASAP!

jbaiera reviewed (approved):
I think once this last fixup is applied, it LGTM!

Once again, sorry about this slipping through the cracks!

jbaiera committed:
Update server/src/main/java/org/elasticsearch/repositories/RepositoryException.java

jbaiera commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into add/hdfs_replication_factory

jbaiera commented:
@elasticmachine generate changelog

jbaiera commented:
@elasticmachine ok to test

jbaiera committed:
Checkstyle

jbaiera committed:
Add changelog file

jbaiera committed:
correct changelog error

jbaiera committed:
Fixup test typing

jbaiera committed:
Appease checkstyle

jbaiera committed:
Precommit why

jbaiera commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into add/hdfs_replication_factory

jbaiera commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into add/hdfs_replication_factory

trakos commented:
Thank you so much for bringing this over the line!
